braindecode/datasets/base.py

```diff
@@ -142,6 +144,88 @@ def transform(self, value):
         self._transform = value
 
 
+class SequenceDataset(WindowsDataset):
+    """Dataset of sequences of consecutive windows.
+
+    Parameters
+    ----------
+    windows : mne.Epochs
+        Windows obtained through the application of a windower to a BaseDataset
+        (see `braindecode.datautil.windowers`).
+    description : dict | pandas.Series | None
+        Holds additional info about the windows.
+    transform : callable | None
+        On-the-fly transform applied to a window before it is returned.
+    seq_len : int
+        Number of consecutive windows in a sequence.
+    step_len : int
+        Number of windows between two consecutive sequences.
+    label_transform : callable | None
+        On-the-fly transform to apply to the sequence of labels. Useful to e.g.
+        select a single label for each sequence.
+    on_missing : 'raise' | 'warn' | 'ignore'
+        What to do if no sequences can be extracted given the number of epochs
+        and sequence parameters. Valid keys are 'raise' | 'warn' | 'ignore'.
+        Default is 'raise'. If on_missing is 'warn' it will proceed but warn,
+        if 'ignore' it will proceed silently.
+    """
+    def __init__(self, windows, description=None, transform=None, seq_len=21,
+                 step_len=1, label_transform=None, on_missing='raise'):
+        super().__init__(windows, description=description, transform=transform)
+
+        if seq_len > len(windows):
+            msg = ('Sequence length is larger than number of windows '
+                   f'({seq_len} > {len(windows)}).')
+            _on_missing(on_missing, msg)
+
+        # Make list of all possible sequences
+        self.seq_len = seq_len
+        self.step_len = step_len
+        self.start_inds = self._compute_seq_inds()
+        self.label_transform = label_transform
+
+    def _compute_seq_inds(self):
+        """Compute sequence indices.
+
+        Returns
+        -------
+        np.ndarray :
+            Array of shape (n_sequences,) containing the indices of the first
+            windows of possible sequences.
+        """
+        return np.arange(
+            0, super().__len__() - self.seq_len + 1, self.step_len)
+
+    def __len__(self):
+        return len(self.start_inds)
+
+    def __getitem__(self, index):
+        """
+        Returns
+        -------
+        np.ndarray :
+            Sequence of windows, of shape (seq_len, n_chs, *).
+        np.array :
+            List of labels of length seq_len, or can be something else if
+            `label_transform` is defined.
+        """
+        start_ind = self.start_inds[index]
+        seq, ys = list(), list()
+        # The following cannot be put into a list comprehension
+        # because of scope requirements for `super()`
+        for i in range(start_ind, start_ind + self.seq_len):
```

Comment:
is this fast enough for our application? Can't we use `__getitem__` to obtain multiple windows? 
I see that this line does not allow this `X = self.windows.get_data(item=index)[0].astype('float32')` but maybe we can modify `WindowsDataset` to make it possible? I think that you can create an additional method that returns multiple windows and use it here and in the `WindowsDataset.__getitem__` where you can select only the first epoch.

Reply:
indeed some performance measurements may be helpful. in case it is fast enough, it does seem nicely decoupled at least this way.